File: blind75/58_invert_binary_tree/invert_binary_tree.go
package invert_binary_tree

// TreeNode represents a node in a binary tree.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// InvertTree inverts a binary tree (mirrors it).
// Uses recursive DFS for O(n) time complexity.
func InvertTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	// Swap left and right children
	root.Left, root.Right = root.Right, root.Left

	// Recursively invert subtrees
	InvertTree(root.Left)
	InvertTree(root.Right)

	return root
}

File: blind75/58_invert_binary_tree/invert_binary_tree.py
from typing import Optional


class TreeNode:
    """Definition for a binary tree node."""

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Inverts a binary tree (mirrors it).
    Uses recursive DFS for O(n) time complexity.

    Args:
        root: Root node of the binary tree

    Returns:
        Root node of the inverted tree
    """
    if root is None:
        return None

    # Swap left and right children
    root.left, root.right = root.right, root.left

    # Recursively invert subtrees
    invert_tree(root.left)
    invert_tree(root.right)

    return root
